### PR TITLE
Allow async functions as router handlers

### PR DESCRIPTION
**Pull request by leoorshansky**
The router currently only allows synchronous functions as callbacks, but one might want to use the asynchronous `fetch()` api within a request handler. This PR adds `AsyncRouter` which takes `|| Box::pin(async { ... })` functions.

Also moved router code into its own `router.rs`

### CODE REVIEW

**Comment by nilslice** [comment]
Thanks, @leoorshansky - I definitely agree we need to support async calls inside the router. Not convinced yet that this is the best API, but it might be the only way with the current setup... let's think about how we might change some other surface area in the router, or the macro layer(?), to enable a different async approach. 

I'll leave this open though for now.

File: rust-sandbox/src/lib.rs [hunk 1]
```diff
@@ -75,7 +75,13 @@ pub async fn main(req: Request) -> Result<Response> {
         )))
     })?;
 
+    let mut async_router = AsyncRouter::new();
+    async_router.get("/", |_req, _params| async {
```
Comment (leoorshansky) [inline rust-sandbox/src/lib.rs:79]: Removed the need for `Box::pin` to be used when adding a handler function, it now looks normal

File: rust-sandbox/src/lib.rs [hunk 2]
```diff
@@ -75,7 +75,13 @@ pub async fn main(req: Request) -> Result<Response> {
...
+        Response::empty()
```
Comment (leoorshansky) [inline rust-sandbox/src/lib.rs:80]: Removed the need for `Box::pin` to be used when adding a handler function, it now looks normal

File: rust-sandbox/src/lib.rs [hunk 3]
```diff
@@ -75,7 +75,13 @@ pub async fn main(req: Request) -> Result<Response> {
...
+    })?;
```
Comment (leoorshansky) [inline rust-sandbox/src/lib.rs:81]: Removed the need for `Box::pin` to be used when adding a handler function, it now looks normal

**Comment by nilslice** [comment]
Nice!

**Pull request merged**